### ISSUE TITLE
e2epod.DeletePodWithWait{,ByName} does not handle pods that get restarted

### ISSUE THREAD

**Issue by danwinship** [kubernetes/kubernetes#126785]
### What happened?

`e2epod.DeletePodWithWaitByName` does (simplified)

```
        err := c.CoreV1().Pods(podNamespace).Delete(ctx, podName, metav1.DeleteOptions{})
        err = WaitForPodNotFoundInNamespace(ctx, c, podName, podNamespace, PodDeleteTimeout)
```

If the pod in question is managed by a controller or operator, then it is possible that it will be restarted in between the Delete and the WaitForPodNotFoundInNamespace, in which case WaitForPodNotFoundInNamespace will wait for the _new_ pod to not be present rather than the old one.

eg, some of the ovn-kubernetes e2e tests use this function to restart ovn-k control-plane pods in various circumstances (the e2e test kills a pod which is managed by a StatefulSet, knowing it will be restarted), and sometimes hits the race condition. eg, [here](https://github.com/ovn-org/ovn-kubernetes/actions/runs/10451772693/job/28939840358?pr=4623):

```
2024-08-19T11:40:56.3501692Z   I0819 11:40:56.349677 79626 delete.go:62] Deleting pod "ovnkube-db-2" in namespace "ovn-kubernetes"
2024-08-19T11:40:56.3606091Z   I0819 11:40:56.360068 79626 delete.go:70] Wait up to 5m0s for pod "ovnkube-db-2" to be fully deleted
...
2024-08-19T11:45:57.4798660Z   [FAILED] failed to delete pod ovnkube-db-2: pod "ovnkube-db-2" was not deleted: expected pod to not be found: Timed out after 300.001s.
2024-08-19T11:45:57.4799987Z   Expected
2024-08-19T11:45:57.4800476Z       <*v1.Pod | 0xc000eba488>: 
2024-08-19T11:45:57.4801044Z           metadata:
2024-08-19T11:45:57.4801863Z             creationTimestamp: "2024-08-19T11:41:00Z"
...
2024-08-19T11:45:57.5299664Z   to be nil
```

note that the `creationTimestamp` of the pod in question is _after_ the deletion.

### What did you expect to happen?

Either

1. DeletePodWithWait{,ByName} checks the `resourceVersion` of the pod before deleting it, and waits specifically for _that_ version of the pod to not exist any more, or
2. DeletePodWithWait{,ByName} are documented as not handling this case

### How can we reproduce it (as minimally and precisely as possible)?

Write an e2e test that deletes a pod that will be recreated automatically (by an `apps` API or by an operator).

### Anything else we need to know?

/sig testing
/area e2e-test-framework

### Kubernetes version

`v1.30.2`

### Cloud provider

N/A

**Comment by danwinship** [comment]
> 1. DeletePodWithWait{,ByName} checks the `resourceVersion` of the pod before deleting it, and waits specifically for _that_ version of the pod to not exist any more

(This would require a new version of `WaitForPodNotFoundInNamespace` as well.)

**Comment by aojea** [comment]
@pohly and @BenTheElder for thoughts

**Comment by danwinship** [comment]
> One alternative for tests which need a solution right now

Yeah, I was hoping there'd be a strong consensus for fixing the current methods in some particular way, and then I could just implement that. But failing that, I can just patch it up in our failing downstream tests. (I'm not aware of any currently-existing k/k e2e tests that are affected by this bug; it's possible that everyone within k/k is only using the functions to delete manually-created pods.)
